src/main/java/io/nats/client/impl/NatsConnection.java

```diff
@@ -944,39 +941,58 @@ String getResponseToken(String responseInbox) {
         return responseInbox.substring(len);
     }
 
-    void cleanResponses(boolean cancelIfRunning) {
+    void cleanResponses(boolean closing) {
         ArrayList<String> toRemove = new ArrayList<>();
 
-        responses.forEach((token, f) -> {
-            if (f.isDone() || cancelIfRunning) {
-                try {
-                    f.cancel(true); // does nothing if already done
-                } catch (CancellationException e) {
-                    // Expected
-                }
+        responsesAwaiting.forEach((token, f) -> {
+            boolean remove = false;
+            if (f.isDone()) {
+                remove = true;
+            }
+            else if (closing) {
+                remove = true;
+                f.cancelClosing();
+            }
+            else if (f.hasExceededTimeout()) {
+                remove = true;
+                f.cancelTimedOut();
+            }
+
+            if (remove) {
                 toRemove.add(token);
                 statistics.decrementOutstandingRequests();
             }
         });
 
         for (String token : toRemove) {
-            responses.remove(token);
+            responsesAwaiting.remove(token);
+        }
+
+        toRemove.clear();
+        responsesRespondedTo.forEach((token, f) -> {
+            if (f.hasExceededTimeout()) {

```

Comment:
now that there is responsesAwaiting and responsesRespondedTo, handle both in cleanup. responsesRespondedTo gets cleaned up after timeout